gateway/websocket_request_context.go

```diff
@@ -0,0 +1,397 @@
+	if err != nil {
+		wrc.updateProtocolObservations(&protocolCtxSetupErrObs)
+		logger.Error().Err(err).Str("endpoint_addr", string(selectedEndpoints[0])).Msg("Failed to build protocol context for websocket endpoint")
+		return fmt.Errorf("failed to build protocol context for websocket endpoint: %w", err)
+	}
+
+	wrc.protocolCtx = protocolCtx
+	logger.Info().Msgf("Successfully built protocol context for websocket endpoint: %s", selectedEndpoints[0])
+
+	return nil
+}
+
+// HandleWebsocketRequest establishes the websocket connection and starts the message handling loop.
+func (wrc *websocketRequestContext) HandleWebsocketRequest(request *http.Request, responseWriter http.ResponseWriter) error {
+	// Create the websocket bridge in the gateway package and start the bridge asynchronously.
+	if err := wrc.initializeWebsocketBridge(request, responseWriter); err != nil {
+		wrc.logger.Warn().Err(err).Msg("Failed to create websocket bridge.")
+		return err
+	}
+
+	// Start the bridge with gateway observations and data reporter
+	// This should be handled asynchronously to avoid blocking
+	go wrc.bridge.StartAsync()
+
+	wrc.logger.Info().Msg("🚀 Websocket bridge started successfully")
+	return nil
+}
+
+// createWebsocketBridge creates a websocket bridge using protocol-specific components.
+// This moves the bridge creation logic from Shannon to the gateway level.
+func (wrc *websocketRequestContext) initializeWebsocketBridge(
+	req *http.Request,
+	w http.ResponseWriter,
+) error {
+	// Hydrate the logger with the websocket bridge specific information.
+	logger := wrc.logger.With("connection_type", "websocket")
+
+	// Get the websocket-specific URL from the selected endpoint.
+	websocketURL, err := wrc.protocolCtx.GetWebsocketEndpointURL()
+	if err != nil {
+		logger.Error().Err(err).Msg("❌ Selected endpoint does not support websocket RPC type")
+		return err
+	}
+	logger = logger.With("websocket_url", websocketURL)
+
+	// Upgrade HTTP request from client to websocket connection.
+	// - Connection is passed to websocket bridge for Client <-> Gateway communication.
+	clientConn, err := websockets.UpgradeClientWebsocketConnection(wrc.logger, req, w)
+	if err != nil {
+		return fmt.Errorf("createWebsocketBridge: %s", err.Error())
+	}
+
+	// Get the headers for the websocket connection.
+	headers := wrc.protocolCtx.GetWebsocketConnectionHeaders()
+
+	// Connect to the endpoint
+	endpointConn, err := websockets.ConnectWebsocketEndpoint(wrc.logger, websocketURL, headers)
+	if err != nil {
+		return fmt.Errorf("createWebsocketBridge: %s", err.Error())
+	}
+
+	// Create protocol-specific message handlers
+	clientHandler := &websocketClientMessageHandler{
+		logger:      logger.With("component", "websocket_client_message_handler"),
+		protocolCtx: wrc.protocolCtx,
+		serviceID:   wrc.serviceID,
+	}
+	endpointHandler := &websocketEndpointMessageHandler{
```

Comment:
This file brings a lot of improvements in general, but "message handler" is still too vague and error-prone.
Please revise to make the continuous flow very clear to the reader.